src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -358,12 +358,134 @@ internal void WriteEntry(EntityDescriptor entityDescriptor, IEnumerable<LinkDesc
+            }
+
+            IEnumerable<ClientPropertyAnnotation> properties;
+
+            properties = entityType.PropertiesToSerialize();
+
+            entry.Properties = this.propertyConverter.PopulateProperties(entityDescriptor.Entity, serverTypeName, properties);
+
+            oDataWriter.WriteStart(entry, entityDescriptor.Entity);
+
+            this.WriteNestedComplexProperties(entityDescriptor.Entity, serverTypeName, properties, oDataWriter);
+            this.WriteNestedResourceInfo(entityDescriptor, bulkUpdateGraph, oDataWriter);
+
+            oDataWriter.WriteEnd(entry, entityDescriptor.Entity);
+        }
+
+        /// <summary>
+        /// Writes a nested resource.
+        /// </summary>
+        /// <param name="entityDescriptor">The entity.</param>
+        /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/>.</param>
+        /// <param name="odataWriter">The ODataWriter used to write the navigation link.</param>
+        internal void WriteNestedResourceInfo(EntityDescriptor entityDescriptor, BulkUpdateGraph bulkUpdateGraph, ODataWriterWrapper odataWriter)
+        {
+            List<Descriptor> relatedDescriptors = bulkUpdateGraph.GetRelatedDescriptors(entityDescriptor);
+            Dictionary<string, List<Descriptor>> groupedRelatedLinks = GroupRelatedLinksByNavigationProperty(relatedDescriptors, null);
+
+            ClientEdmModel model = this.requestInfo.Model;
+            ClientTypeAnnotation clientType = model.GetClientTypeAnnotation(model.GetOrCreateEdmType(entityDescriptor.Entity.GetType()));
+
+            foreach (KeyValuePair<string, List<Descriptor>> relatedLinks in groupedRelatedLinks)
+            {
+                LinkDescriptor relatedEnd = null;
+                ODataNestedResourceInfo navigationLink = new ODataNestedResourceInfo();
+
+                //these items will all be of the same type ~ we'll be using the
+                //first item in determining the type of the descriptor.
+                Descriptor item = relatedLinks.Value.FirstOrDefault();
+
+                if (item is EntityDescriptor entDescriptor)
+                {
+                    relatedEnd = new LinkDescriptor(entDescriptor.ParentEntity, entDescriptor.ParentProperty, entDescriptor.Entity, this.requestInfo.Model);
+                }
+                else
+                {
+                    relatedEnd = item as LinkDescriptor;
+                }
+
+                navigationLink.Name = relatedEnd.SourceProperty;
+                bool isCollection = clientType.GetProperty(relatedEnd.SourceProperty, UndeclaredPropertyBehavior.ThrowException).IsEntityCollection;
+
+                odataWriter.WriteStart(navigationLink, relatedEnd.Source, relatedEnd.Target);
+
+
+                if (isCollection)
+                {
+                    WriteResourceSet(relatedLinks.Value, bulkUpdateGraph, odataWriter);
+                }
+                else
+                {

```

Comment:
Which scenario leads to this code block being executed? Is there a test I can run to get this block executed?

Reply:
Single value navigation property